website/cue/reference/remap/functions/encode_gzip.cue

```diff
@@ -0,0 +1,37 @@
+package metadata
+
+remap: functions: encode_gzip: {
+	category:    "Codec"
+	description: """
+		Encodes the `value` to [Gzip](\(urls.gzip)).
+		"""
+
+	arguments: [
+		{
+			name:        "value"
+			description: "The string to encode."
+			required:    true
+			type: ["string"]
+		},
+		{
+			name:        "compression_level"
+			description: "The default compression level."
+			required:    false
+			type: ["integer"]
+			default: 6
+		},
+	]
+	internal_failure_reasons: []
```

Comment:
I suspect this can't actually fail in practice since we are passing in a `Vec` to read into but that the `io::Read` trait has to return errors incase the underlying stream actually can return an error. @zamazan4ik I'd suggest you verify that. If it is true, we can make this infallible and add an `expect` to the encode call in `encode_gzip.rs` (maybe with some explanatory comments).

Reply:
That would mirror the `base` encoders IIRC as well, decoding is fallible because the input could be non-sensical (for that encoding) while encoding is infallible.

I definitely just checked that the underlying function returned a `Result` and figured fallible was reasonable.

Reply:
Yes, it returns the result so that is why I've marked it as fallible. However, I do not know right now the exact reasons, in that exact cases the `encode` function could fail. Maybe, I need to dig a bit more into the `encode` internals.

Reply:
👍 I believe this is just because the `io::Read` trait we are using here has to return errors since it is used with types that can return errors. I just don't think that `Vec` can actually return an error here. It'll require some digging but if it is true would make the interface better for users rather than requiring them to handle an error that would never happen in practice.

Reply:
Well, the corresponding method inside `flate2` says the following: https://doc.rust-lang.org/nightly/std/io/trait.Read.html#tymethod.read

Honestly, not sure, what is useful for the users we could put into the Vector documentation for `encode` function.

Reply:
That's the docs for the `io::Read` trait 🙂  That trait has to allow for failure even if the type implementing it can never fail to read (e.g. `Vec` or `Bytes` will never return an error when implementing that trait).

You'd want to look at the trait implementation: https://docs.rs/flate2/latest/src/flate2/gz/write.rs.html#156-160

This calls `read` on whatever the `inner_mut()` returns: https://github.com/rust-lang/flate2-rs/blob/6d8a1fdaf3bd5a5e272d4c2825b48884bd5c72ad/src/gz/write.rs#L71-L73

Which calls https://github.com/rust-lang/flate2-rs/blob/6d8a1fdaf3bd5a5e272d4c2825b48884bd5c72ad/src/zio.rs#L189-L191

In this case the `obj` is what we passed in, `Bytes`, whose `read` method can never fail: https://docs.rs/bytes/latest/src/bytes/buf/reader.rs.html#66-71

Even though the return signature needs to allow for failures to satisfy the `io::Read` trait.

The `write` also needs to never fail. In this case we are passing a `Vec` which will never fail to `write`: https://doc.rust-lang.org/src/std/io/impls.rs.html#383-386

Admittedly this sort of manual validation is brittle since if we passed in a different type to read from, that could fail, then this function could fail too. I still think it is worth it to improve the UX and not require VRL users to handle failures where none can occur. Adding sufficient code comments should cause us to reconsider the conditions were any of the types to change to types that could fail to read/write (unlikely).

Let me know if that was clear!

Reply:
> I believe this is just because the io::Read trait we are using here has to return errors since it is used with types that can return errors. I just don't think that Vec can actually return an error here. It'll require some digging but if it is true would make the interface better for users rather than requiring them to handle an error that would never happen in practice.

Hm, that makes sense. Changed to infallible.

Reply:
Fixed.